### DISCUSSION TITLE
How to detect when somebody close a browser tab in Next.js?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#63148] [Help] [answered]
### Summary

This is an old topic that I had open last year, and apparently other people had the same topic, but I didn't see any solutions to it.

I understand that NEXT.JS doesn't have a window for the browser, so, how would Next.js detect an event coming from closing a tab on the window?

My application in specific would be triggering a function to send email history of a chat when the user closes the window/tab/exit the website.

So far I have been working on sending email history on a button click that closes the chat `div` (which kind of works),  but nothing similar to shutting down the tab on the browser.

I was thinking... if I can't make it work in Next.js, maybe I would need to create a separate website (using REACT maybe?) where I can get send information on tab closing just for the chat bot...  but then I have been lazy on duplicating all that code to put it somewhere else.

Are there any recent solutions from Next.js  on how to trigger a function on a tab close event? 

### Additional information

```js
If you want to see the entire code for the page sending an email history for the chat it would be this one...

Line 88 is where I send the email out on a close button for the chat display. 

https://github.com/Elindo586/nextprojectb/blob/main/components/chat/message-input.js
```

**Comment by icyJoseph** [comment]
Hi,

This has always been possible, by just attaching event listeners to the window within a use-effect callback.

- https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event#examples

Though if you google around beforeunload is also mentioned.

Regardless, the solution has always been to just attach an event listener to the window, within a use-effect.

**Reply by icyJoseph** [reply]
So you just need to call `emailChatHistory`, when either event above fires, like this:

```jsx
  useEffect(() => {
    document.addEventListener("visibilitychange", () => {
      if (document.visibilityState === "hidden") {
        if (chatToServer.length > 0) {
          emailChatHistory();
        }
      }
    });
  });

```

This should do the trick, if it doesn't try first with `beforeunload`. You should also keepalive to `emailChatHistory`:

- https://developer.mozilla.org/en-US/docs/Web/API/fetch#keepalive

```js
 const response = await fetch("/api/openai", {
      method: "POST",
      headers: {
        Accept: "application/json, text/plain, */*",
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
      keepalive: true
    });
```

**Reply by Elindo586** [reply]
Let me try it... I think last time I didn't keepalive the api...

**Reply by Elindo586** [reply]
Ok... this works if you close the tab, yes.... it also works if you hide the tab or navigate to another page, which can send more emails than needed.

Right now the chat is created inside a component that shows on every page, but I could make it on its own page and fix some of the navigation problems of sending email, but it would still send emails every time the tab is hidden.

Ideally I would want to send an email only when the tab/browser is closed.

**Reply by icyJoseph** [reply]
Well, try with the beforeunload event, does that work?

**Reply by Elindo586** [reply]
I am gonna try beforeunload tomorrow. I know it doesn't work all that good with some browsers, but gonna try it. 

I added a little extra code on the `visibilitychange`  this time, and it worked on tab close...  Last time I used it wasn't working on tab close but now it works.... just a bit annoying with so many extra emails.

**Reply by Elindo586** [reply, marked as answer]
Ok.. this works, even if it isn't working for every browser...  but it does most of the job in combination with the other settings as well...

 ```
useEffect(() => {
    window.addEventListener("beforeunload", () => {
      if (chatToEmailServer.chatToServer.length > 0) {
        emailChatHistory();
        setFirstMsg(true);
      }
    });
  }, [chatToEmailServer.chatToServer, firstMsg]);
```
